linux/bindings/src/thread.rs

```diff
@@ -0,0 +1,266 @@
+        T: Send + 'static,
+    {
+        // SAFETY: `ThreadItem` is initialized by `pin_init`, it is valid, and it
+        // should not be dropped until the kthread stops.
+        let item_ptr = unsafe {
+            let pinned_item = Box::pin_init(ThreadItem::try_new(closure))?;
+            Box::into_raw(Pin::into_inner_unchecked(pinned_item))
+        };
+
+        // SAFETY: both `threadfn` and `item_ptr` are valid, and `name` will be
+        // checked by `format_name`.
+        let task_ptr = unsafe {
+            crate::kthread_create_on_node(
+                Some(Self::threadfn::<T>),
+                item_ptr as _,
+                crate::NUMA_NO_NODE,
+                Self::format_name(name).as_ptr() as _,
+            )
+        };
+
+        // SAFETY: no safety requirement on this FFI call.
+        if unsafe { kernel::bindings::IS_ERR(task_ptr as _) } {
+            // SAFETY: `item_ptr` is initialized above, and it should be dropped
+            // here due to the kthread creation failure.
+            unsafe { Box::from_raw(item_ptr) };
+            return Err(ENOMEM);
+        }
+
+        // SAFETY: `task_ptr` is a valid `task_struct`, we get a reference
+        // to it, so that it can't go away util the kthread stops.
+        unsafe {
+            crate::get_task_struct(task_ptr);
+            crate::wake_up_process(task_ptr);
+        }
+
+        // SAFETY: it is checked above that `task_ptr` is valid.
+        Ok(Self {
+            inner: unsafe { NonNull::new_unchecked(task_ptr) },
+        })
+    }
+
+    /// Returns true if the user `closure` has been finished.
+    fn is_finished<T>(&self) -> bool {
+        // SAFETY: the `self.inner` is a valid `task_struct`, `kthread_data`
+        // returns value specified on kthread creation, which is a valid
+        // `ThreadItem`.
+        unsafe {
+            let item_ptr = crate::kthread_data(self.inner.as_ptr()) as *mut ThreadItem<T>;
+            (*item_ptr).is_finished()
+        }
+    }
+
+    /// Waits the kthread to finish and retrieves the value of `ThreadItem`.
+    ///
+    /// # Safety
+    ///
+    /// This method must be called, and only be called once, since we should drop
+    /// `ThreadItem` here. And after this, it should not be used ever.
+    unsafe fn stop<T>(&self) -> T {
+        // SAFETY: the `self.inner` is a valid `task_struct`, so the `kthread_data`
+        // is also a valid `ThreadItem`.
+        let item = unsafe {
+            let item_ptr = crate::kthread_data(self.inner.as_ptr()) as *mut ThreadItem<T>;
+            Box::from_raw(item_ptr)
+        };
+
+        // Wait `ThreadItem` to be executed and take the value.
+        item.wait_value()
+    }
+}
+
+impl Drop for Thread {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `task_struct`, we get a reference
+        // to it on kthread creation, now put it here.
+        unsafe { crate::put_task_struct(self.inner.as_ptr()) };
```

Comment:
Memory management of `task_struct` seems fine because:
- `get_task_struct()` and `put_task_struct()` come in pair, in `Thread::create()` and `Thread::drop()`, individually, which are both private fns and can only be called once .
- Since `Thread` is a field in `JoinHandle`, `JoinHandle::drop()` would be called ahead of `Thread::drop()`, which means 1. `Thread::stop()`, which used to get kthread data `ThreadItem` and free it, is called ahead of `put_task_struct()`, which leaves freeing `task_struct` at last. The logic seems reasonable.